### PR TITLE
Restore TOX_SKIP_ENV filtering

### LINKED ISSUE #2698: tox 4 ignores TOX_SKIP_ENV
## Issue

My `make release` for [gitlab-trace](https://github.com/mgedmin/gitlab-trace) runs `TOX_SKIP_ENV=check-manifest tox -p auto`.  This worked with tox 3.x.  It now fails in tox 4, because it doesn't skip the check-manifest toxenv.

Since the documentation page describing the TOX_SKIP_ENV has 'latest' in the URL (https://tox.wiki/en/latest/config.html#envlist), I think it should still be supported?

## Environment

Provide at least:

- OS: Ubuntu 22.10
- `pip list` of the host Python where `tox` is installed:

```console
$ pipx runpip tox list
Package       Version
------------- -------
cachetools    5.2.0
chardet       5.1.0
colorama      0.4.6
distlib       0.3.6
filelock      3.8.2
packaging     22.0
pip           22.3.1
platformdirs  2.6.0
pluggy        1.0.0
py            1.11.0
pyparsing     3.0.9
pyproject_api 1.2.1
setuptools    65.6.3
six           1.16.0
toml          0.10.2
tomli         2.0.1
tox           4.0.2
virtualenv    20.17.1
wheel         0.38.4
```

### PR DESCRIPTION
**Pull request by mgedmin**
This is still documented in docs/config.rst.  The only missing thing is the reporting at verbosity level 2.

I feel bad about the missing reporting.  Could someone help me figure it out?

Fixes #2698.

# Thanks for contribution

Please, make sure you address all the checklists (for details on how see
[development documentation](http://tox.readthedocs.org/en/latest/development.html#development))!

- [x] ran the linter to address style issues (`tox -e fix`)
- [x] wrote descriptive pull request text
- [x] ensured there are test(s) validating the fix
- [x] added news fragment in `docs/changelog` folder
- [x] updated/extended the documentation

### CODE REVIEW

**Comment by mgedmin** [comment]
I am very confused about logging levels.  `tox --help` says

```
verbosity:
  every -v increases, every -q decreases verbosity level, default INFO, map 0=CRITICAL|1=ERROR|2=WARNING|3=INFO|4=DEBUG|5=NOTSET

  -v, --verbose               increase verbosity (default: 2)
  -q, --quiet                 decrease verbosity (default: 0)
```

so level 2 is WARNING, and the default level is INFO, which is 3, but also the default level is 2, which is WARNING.

Anyway, I'm logging with .info(), and it's visible only if you pass a single -v to tox, which means the actual default is WARNING and the --help output lies to me?

Should I change the log level of this message to .warning() so it would match docs/config.rst?

File: docs/changelog/2698.bugfix.rst [hunk 1]
```diff
@@ -0,0 +1,2 @@
+``TOX_SKIP_ENV`` environment variable now works again (it was unintentionally
```
Comment (gaborbernat) [inline docs/changelog/2698.bugfix.rst:1]: Please wrap at 120 characters.

File: tests/session/test_env_select.py [hunk 2]
```diff
@@ -70,3 +70,27 @@ def test_factor_select(tox_project: ToxProjectCreator) -> None:
...
     outcome.assert_success()
     outcome.assert_out_err("py310-django20-cov\npy39-django20-cov\n", "")
+
+
+def test_tox_skip_env(tox_project: ToxProjectCreator, monkeypatch: MonkeyPatch) -> None:
+    monkeypatch.setenv("TOX_SKIP_ENV", "m[y]py")
+    ini = """
+        [tox]
+        env_list = py3{10,9},mypy
+        """
```
Comment (gaborbernat) [inline tests/session/test_env_select.py:80]: You can include this in one line considering its simplicity.

File: tests/session/test_env_select.py [hunk 3]
```diff
@@ -70,3 +70,27 @@ def test_factor_select(tox_project: ToxProjectCreator) -> None:
...
+    outcome = project.run("l", "--no-desc", "-q")
+    outcome.assert_success()
+    outcome.assert_out_err("py310\npy39\n", "")
+
+
+def test_tox_skip_env_logs(tox_project: ToxProjectCreator, monkeypatch: MonkeyPatch) -> None:
+    monkeypatch.setenv("TOX_SKIP_ENV", "m[y]py")
+    ini = """
+        [tox]
+        env_list = py3{10,9},mypy
```
Comment (gaborbernat) [inline tests/session/test_env_select.py:91]: You can include this in one line considering its simplicity.

File: src/tox/session/env_select.py [hunk 4]
```diff
@@ -101,11 +107,15 @@ class _ToxEnvInfo:
...
 class EnvSelector:
+
+    _warned_about: set[str]  #: shared set of skipped environments that were already warned about
+
     def __init__(self, state: State) -> None:
         # needs core to load the default tox environment list
         # to load the package environments of a run environments we need the run environment builder
         # to load labels we need core + the run environment
         self.on_empty_fallback_py = True
+        self._warned_about = set()
```
Comment (gaborbernat) [inline src/tox/session/env_select.py:118]: You can move line 111 here; no need to define and then initialize it here; just initializing it is enough

File: src/tox/session/env_select.py [hunk 5]
```diff
@@ -320,14 +330,19 @@ def iter(
 
         :return: an iteration of tox environments
         """
-        ignore_envs: set[str] = set()
+        tox_env_filter = os.environ.get("TOX_SKIP_ENV")
+        tox_env_filter_re = re.compile(tox_env_filter) if tox_env_filter is not None else None
```
Comment (gaborbernat) [inline src/tox/session/env_select.py:334]: This definition should happen in the constructor of this class, not every time on this call, no?

File: docs/changelog/2698.bugfix.rst (hunk 1 above)
Comment (gaborbernat) [reply docs/changelog/2698.bugfix.rst:1]: Also do you want to add your user name here? 

```
- :by:`x`.
```

**Review by gaborbernat — CHANGES_REQUESTED** [review]

**Comment by mgedmin** [comment]
Ah, sorry, I meant to take another look at this but was too busy yesterday.

Thank you for finishing it!

File: src/tox/session/env_select.py [hunk 6]
```diff
@@ -88,6 +93,8 @@ def register_env_select_flags(
         add_to.add_argument("-m", dest="labels", metavar="label", help=help_msg, default=[], type=str, nargs="+")
         help_msg = "factors to evaluate"
         add_to.add_argument("-f", dest="factors", metavar="factor", help=help_msg, default=[], type=str, nargs="+")
+    help_msg = "exclude all environments selected that match this regular expression"
+    add_to.add_argument("--skip-env", dest="skip_env", metavar="re", help=help_msg, default="", type=str)
```
Comment (mgedmin) [inline src/tox/session/env_select.py:97]: I like this even more than the environment variable!

**Pull request merged**